### PR TITLE
New logger

### PR DESCRIPTION
**Pull request by dagar**
continuation of #2780

@DrTon @againagainst @kd0aij @LorenzMeier - Anyone interested in finishing this?

TODO:
- [ ] fix ROS
- [ ] SITL test
- [ ] nuttx test
- [x] increase message size? (currently 254)
- [ ] higher rate for EKF2 replay?

### CODE REVIEW

File: ROMFS/px4fmu_common/init.d/rc.logging [hunk 1]
```diff
@@ -7,18 +7,18 @@ if [ -d /fs/microsd ]
...
-		if sdlog2 start -r 40 -a -b 3 -t
+		if logger start -r 40
 		then
 		fi
 	else
 		# check if we should increase logging rate for ekf2 replay message logging
 		if param greater EKF2_REC_RPL 0
 		then
-			if sdlog2 start -r 500 -e -b 20 -t
+			if logger start -r 500
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d/rc.logging]: TODO: test EKF2_REC_RPL

**Comment by kd0aij** [comment]
@dagar @jgoppert Thanks, I was able to work with #4041 
But it looks like a few things need to be done to get the logger working (message and format string size to start with).

**Comment by kd0aij** [comment]
The system arms, but start_log is never called because _vehicle_status_sub.get().arming_state remains zero.

**Comment by kd0aij** [comment]
Can anyone tell me why this error might be occurring now, and how to work around it?

log_writer: can't open log file /fs/microsd/log/sess020/log001.ulg: Too many open files

**Comment by LorenzMeier** [comment]
@kd0aij that means that the app is doing one of two things: it's leaking file descriptors, or its opening more topics than the maximum number of file descriptors allowed per process. Most likely it needs to be smarter about which topics to log and not all by default.

**Comment by kd0aij** [comment]
@LorenzMeier Thanks, reducing the number of topics resulted in a log file with some data. Progress!

**Comment by kd0aij** [comment]
Although "_vehicle_status_sub.get().arming_state" compiles, it always returns zero, instead of the actual value

**Comment by kd0aij** [comment]
There are errors in the autogenerated cpp files e.g. sensor_baro.cpp contains 
```
const char *__orb_sensor_baro_fields = "SENSOR_BARO:float pressure;float altitude;float temperature;uint64_t error_count;";
```
but the correct field order is:
```
const char *__orb_sensor_baro_fields = "SENSOR_BARO:uint64_t error_count;float pressure;float altitude;float temperature;";
```

**Comment by kd0aij** [comment]
The cpp header files for autogenerated messages say "Auto-generated by genmsg_cpp", but there is no such file in the source tree. What is the real answer to that question?

**Comment by dagar** [comment]
It doesn't have the rest of this? 

    /* Auto-generated by genmsg_cpp from file @file_name_in */

https://github.com/PX4/Firmware/pull/4045/files#diff-90e94c514f9b3676e1b4ccafe737626aR50

**Comment by dagar** [comment]
If you build posix_sitl_default with clang you can catch some of this sizing stuff locally.

**Comment by LorenzMeier** [comment]
@jgoppert Can you help @kd0aij With the field ordering of the log messages in Python? Essentially the issue is that meta data and fields need to agree.

**Comment by kd0aij** [comment]
When I said there was no "genmsg_cpp" in the source tree, I meant that the comment is useless: I still don't know what code generated the broken cpp sources.

**Comment by LorenzMeier** [comment]
@kd0aij Its in Tools/genmsg

**Comment by kd0aij** [comment]
I was afraid of that. Is there any relation to this: http://docs.ros.org/api/genmsg/html/

**Comment by LorenzMeier** [comment]
@kd0aij Its just a couple of python files for now. It doesn't really matter who else is using it. Are all fields re-sorted into their opposite order? If yes, writing out the meta info in the same order would be all it takes to fix the issue, if I understand it correctly.

**Comment by kd0aij** [comment]
@LorenzMeier did you mean Tools/gencpp instead of Tools/genmsg?
In any case it's not at all clear why the generated .h and input .msg files disagree on the order of fields. It does appear that the format string in the generated .cpp file has the same field order as the .msg file. 
I found a "msg.h.template" in the gencpp folder, but without documentation it's pretty hard to guess what's going on.

**Comment by kd0aij** [comment]
getting a small number of dropouts with the code currently checked in: logging accel, gyro, attitude and status topics at every ORB update (250Hz). throughput 74K bytes/sec with 32K byte buffer:
'''
Processes: 24 total, 3 running, 21 sleeping
CPU usage: 34.00% tasks, 0.60% sched, 65.40% idle
Uptime: 163.627s total, 114.611s idle

 PID COMMAND                   CPU(ms) CPU(%)  USED/STACK PRIO(BASE) STATE 
   0 Idle Task                  114611 65.398     0/    0   0 (  0)  READY 
   1 hpwork                       7153  4.548   776/ 1592 192 (192)  w:sig 
   2 lpwork                        586  0.372   356/ 1592  50 ( 50)  w:sig 
   3 init                         1614  0.000  1316/ 2496 100 (100)  w:sem 
 199 log_writer                   1343  1.491   692/ 1016  60 ( 60)  w:sem 
  84 dataman                         1  0.000   652/ 1192  90 ( 90)  w:sem 
 105 sensors                      3783  2.237  1660/ 1992 250 (250)  w:sem 
 107 gps                            70  0.000   740/ 1192 220 (220)  w:sem 
 109 commander                     767  0.447  3116/ 3592 140 (140)  w:sig 
 110 commander_low_prio              5  0.000   596/ 2872  50 ( 50)  w:sem 
 117 mavlink_if0                  1463  0.894  2324/ 2696 100 (100)  w:sig 
 118 mavlink_rcv_if0                11  0.000   828/ 2096 175 (175)  w:sem 
 124 mavlink_if1                  1206  0.745  2332/ 2696 100 (100)  w:sig 
 125 mavlink_rcv_if1                13  0.000   828/ 2096 175 (175)  w:sem 
 158 attitude_estimator_q        11228  6.935  2092/ 2496 250 (250)  w:sem 
 164 position_estimator_inav      3755  2.386  4764/ 5296 250 (250)  w:sem 
 166 mc_att_control               3475  2.162  1148/ 1496 250 (250)  w:sem 
 168 mc_pos_control                994  0.596  1068/ 1896 250 (250)  w:sem 
 175 navigator                       7  0.000   820/ 1496 105 (105)  w:sem 
 183 frsky_telemetry                 2  0.000   564/ 1096 200 (200)  w:sem 
 186 mavlink_if2                  7831  4.996  2316/ 2696 100 (100)  w:sig 
 187 mavlink_rcv_if2                13  0.000   828/ 2096 175 (175)  w:sem 
 197 logger                       2091  2.237  2708/ 3096 250 (250)  w:sig 
 204 top                           679  3.952  1228/ 1696 100 (100)  RUN   
nsh> 
logger:     74.2 Kbytes/sec, 12407 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 6542 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 12122 highWater,  0 dropouts, 0.000 sec max, free heap: 19664
logger:     74.2 Kbytes/sec, 11088 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 11519 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger: dropout, available: 32708/32768
logger: dropout length: 0.091 seconds
logger: dropout, available: 32766/32768
logger: dropout length: 0.039 seconds
perf
sd fsync: 26 events, 0 overruns, 1357857us elapsed, 52225us avg, min 9408us max 93910us 35985.941us rms
sd write: 2675 events, 0 overruns, 44871187us elapsed, 16774us avg, min 11us max 472900us 13206.652us rms
nsh> 
logger:     73.3 Kbytes/sec, 21930 highWater,  2 dropouts, 0.091 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 10949 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 11087 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 11837 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 11837 highWater,  0 dropouts, 0.000 sec max, free heap: 19664
logger: dropout, available: 32634/32768
logger: dropout length: 0.009 seconds
logger: dropout, available: 32740/32768
logger: dropout length: 0.057 seconds
logger:     73.7 Kbytes/sec, 25629 highWater,  2 dropouts, 0.057 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 12122 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 10949 highWater,  0 dropouts, 0.000 sec max, free heap: 19664
logger:     74.2 Kbytes/sec, 11552 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 11267 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 10982 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 11267 highWater,  0 dropouts, 0.000 sec max, free heap: 19664
logger:     74.2 Kbytes/sec, 11837 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
logger:     74.2 Kbytes/sec, 10949 highWater,  0 dropouts, 0.000 sec max, free heap: 19040
logger:     74.2 Kbytes/sec, 11120 highWater,  0 dropouts, 0.000 sec max, free heap: 19664
logger:     74.2 Kbytes/sec, 10949 highWater,  0 dropouts, 0.000 sec max, free heap: 19552
commander disarm
nsh> logger: stop log
log_writer: stopped, bytes written: 22903952

nsh> free
             total       used       free    largest
Mem:        227840     208128      19712      16928
```

**Comment by kd0aij** [comment]
Just looked at the spec again (https://www.sdcard.org/downloads/pls/pdf/part1_410.pdf). We need to provide a 250 msec buffer to avoid dropouts, for all SDHC speed classes. Since that's the spec, it's unwise to rely on shorter buffers.

> There are two types of busies in a multiple block write operation.
> (1) Write busy at block gap (without CMD12) is maximum 250ms
> (2) Write busy after CMD12 is maximum 250ms (500ms for SDXC)

Since I'm seeing the write function block for nearly 500msec, there's probably something wrong with the driver, assuming that my SDHC cards (Kingston class 4 and Samsung class 10) meet the spec. In any case, a throughput of only 100K bytes/sec will require a >25 KByte buffer to guarantee that no data is dropped. Perhaps we should consider adding a 1Mb serial SRAM to use as a buffer. Given my observations of 500msec blocking intervals, that would be sufficient to maintain throughput of 250KB/sec without drops. If we find and eliminate the extra 250msec delay in the driver, we could have 500KB/sec throughput. CPU load doesn't seem to be a problem; the writer thread is below 2% and the total logger usage is roughly 4% at 74KB/sec. Extrapolating the writer load to 250KBps, it would be about 6%.

**Comment by LorenzMeier** [comment]
@jgoppert Would you mind connecting to @kd0aij and maybe having a look at the Python generation? This would result in a much better logger, and all what is missing seems to be right now some field sorting in Python.

**Comment by kd0aij** [comment]
@jgoppert It looks like the problem may be the sorting in print_parsed_fields() of msg/templates/uorb/msg.h.template

**Comment by kd0aij** [comment]
removing the sort from msg.h.template causes hardfault in commander. 
Adding the sort to the python code generating the format string (instead) is **working OK**. 

Will need to add more code if we want to set different update intervals for each topic.

**Comment by LorenzMeier** [comment]
Awesome! Need to re-read the diff to comment on topic listing.

**Comment by LorenzMeier** [comment]
@kd0aij This approach will help you to only spend an uORB handle for topics which are actually updating:
https://github.com/PX4/Firmware/blob/master/src/modules/sdlog2/sdlog2.c#L901-L923

File: ROMFS/px4fmu_common/init.d/rc.logging (hunk 1 above)
Comment (LorenzMeier) [inline ROMFS/px4fmu_common/init.d/rc.logging]: We will phase out the replay stuff

**Comment by dagar** [comment]
continued here - https://github.com/PX4/Firmware/pull/4117

**Pull request closed without merging**